gobblin-modules/gobblin-gitter/src/main/java/org/apache/gobblin/source/GitterMessageExtractor.java

```diff
@@ -0,0 +1,270 @@
+          JsonArray roomsResponse = parser.parse(roomsResponseStr).getAsJsonArray();
+
+          for (JsonElement roomResponse : roomsResponse) {
+            // Name should always be present, so NPE should happen if well-formed response
+            if (this.gitterRoom.equals(roomResponse.getAsJsonObject().get("name").getAsString())) {
+              this.gitterRoomId = roomResponse.getAsJsonObject().get("id").getAsString();
+              break;
+            }
+          }
+          if (null == this.gitterRoomId) {
+            log.error(String.format("Could not fetch Room Id for specified room: %s "
+                + "Room not found in response: %s", this.gitterRoom, roomsResponseStr));
+            throw new RuntimeException(String.format("Could not fetch Room Id for specified room: %s "
+                + "Room not found in response: %s", this.gitterRoom, roomsResponseStr));
+          }
+          log.info(String.format("Obtained Gitter message room id: %s", this.gitterRoomId));
+          break;
+        }
+        default: {
+          log.error(String.format("Could not fetch Room Id for specified room: %s "
+              + "Got response code: %s", this.gitterRoom, responseCode));
+          throw new RuntimeException(String.format("Could not fetch Room Id for specified room: %s "
+              + "Got response code: %s", this.gitterRoom, responseCode));
+        }
+      }
+    } catch (IOException | ParseException ex) {
+      log.error(String.format("Could not fetch Room Id for specified room: %s ", this.gitterRoom), ex);
+      throw new RuntimeException(ex);
+    }
+  }
+
+  @Override
+  public Schema getSchema()
+      throws IOException {
+    // Use schema from config if provided, else use default Gitter schema
+    String schemaString = IOUtils.toString(this.getClass().getResourceAsStream("/gitter/message.avsc"),
+        StandardCharsets.UTF_8);
+    this.gitterMessageSchema = new Schema.Parser().parse(workUnit.getProp(ConfigurationKeys.SOURCE_SCHEMA, schemaString));
+    return this.gitterMessageSchema;
+  }
+
+  @Nullable
+  @Override
+  public GenericRecord readRecord(GenericRecord reuse)
+      throws DataRecordException, IOException {
+    if (gitterMessages.size() == 0) {
+      // We fetch Gitter messages in batches / pages because of two reasons:
+      // 1. There is no way to easily figure out id for next message to pull, you can only pull in batch reliably.
+      // 2. It is anyway optimal (lesser HTTP calls) to pull in batches.
+      fetchGitterMessages(this.gitterMessages);
+      if (gitterMessages.size() == 0) {
+        // No more messages left to fetch, finish
+        return null;
+      }
+    }
+
+    // Keep moving rolling watermark for messages consumed, so that next time pagination hits
+    // .. we continue from there.
+    GenericRecord message = this.gitterMessages.poll();
+    this.gitterRollingWatermark = new BigInteger(message.get("id").toString(), 16);
```

Comment:
What does 16 mean here? Is it 16 bytes for BigInteger? Is it possible to use macro? Or if we know the id format, like bigInteger lenghth, can we get a pre-defined length?